src/deploy/functions/cel.ts

```diff
@@ -0,0 +1,345 @@
+  return typeof value === "string" && value.includes("{{") && value.includes("}}");
+}
+function isIdentityExpression(value: CelExpression): value is IdentityExpression {
+  return identityRegexp.test(value);
+}
+function isEqualityExpression(value: CelExpression): value is EqualityExpression {
+  return equalityRegexp.test(value);
+}
+function isDualEqualityExpression(value: CelExpression): value is DualEqualityExpression {
+  return dualEqualityRegexp.test(value);
+}
+function isTernaryExpression(value: CelExpression): value is TernaryExpression {
+  return ternaryRegexp.test(value);
+}
+function isLiteralTernaryExpression(value: CelExpression): value is LiteralTernaryExpression {
+  return literalTernaryRegexp.test(value);
+}
+function isDualTernaryExpression(value: CelExpression): value is DualTernaryExpression {
+  return dualTernaryRegexp.test(value);
+}
+
+export class ExprParseError extends FirebaseError {}
+
+/**
+ * Resolves a CEL expression of a supported form, guaranteeing the provided primitive type:
+ * - {{ params.foo }}
+ * - {{ params.foo == 24 }}
+ * - {{ params.foo == params.bar }}
+ * - {{ params.foo == 24 ? "asdf" : params.jkl }}
+ * - {{ params.foo == params.bar ? "asdf" : params.jkl }}
+ * - {{ params.foo ? "asdf" : params.jkl }}, when foo is of boolean type
+ * Values interpolated from params retain their type defined in the param;
+ * it is an error to provide a CEL expression that coerces param types
+ * (i.e testing equality between a IntParam and a BooleanParam). It is also
+ * an error to provide a CEL expression that evaluates to a value of a type
+ * other than provided as wantType.
+ */
+export function resolveExpression(
+  wantType: L,
+  expr: CelExpression,
+  params: Record<string, ParamValue>
+): Literal {
+  // N.B: Since some of these regexps are supersets of others--anything that is
+  // params\.(\S+) is also (.+)--the order in which they are tested matters
+  if (isIdentityExpression(expr)) {
+    return resolveIdentity(wantType, expr, params);
+  } else if (isDualTernaryExpression(expr)) {
+    return resolveDualTernary(wantType, expr, params);
+  } else if (isLiteralTernaryExpression(expr)) {
+    return resolveLiteralTernary(wantType, expr, params);
+  } else if (isTernaryExpression(expr)) {
+    return resolveTernary(wantType, expr, params);
+  } else if (isDualEqualityExpression(expr)) {
+    return resolveDualEquality(expr, params);
+  } else if (isEqualityExpression(expr)) {
+    return resolveEquality(expr, params);
+  } else {
+    throw new ExprParseError("CEL expression '" + expr + "' is of an unsupported form");
+  }
+}
+
+function assertType(wantType: L, paramName: string, paramValue: ParamValue) {
+  if (
+    (wantType === "string" && !paramValue.legalString) ||
+    (wantType === "number" && !paramValue.legalNumber) ||
+    (wantType === "boolean" && !paramValue.legalBoolean)
+  ) {
+    throw new ExprParseError(`illegal type coercion of param ${paramName} to type ${wantType}`);
```

Comment:
nit* I think all FirebaseError messages are expected to be read by the end user and hence start with capital letters (like a sentence) rather than being an error message (like go)/

Reply:
Got it. That makes sense.